### DISCUSSION TITLE
How to make your own match files

### DISCUSSION THREAD

**Discussion by sildater** [CPJKU/partitura#324] [HOWTO]
Match files encode a note-wise alignment of a score and a performance.
To create one, you need:
- A score
- A performance
- An alignment*

*If you don't have an alignment, there are of course automated tools for this (which produce the occasional error). We have a library for this on [PyPI](https://github.com/sildater/parangonar) or on [github](https://github.com/sildater/parangonar). In both links there are the same introductory code examples, the second of which should cover your use case. For automatic alignment to work well you need to check two things:
- the musical material you want to align has unique note Ids. If your musicxml file doesn't contain any Ids you can generate them via `pt.load_musicxml(..., force_ids=True)
- the musical material in both scores has to match roughly, that is, there aren't any large mismatches: full repeats not played, extra sections, an entire coda, etc. anything that stretches multiple measures can be a problem. If your musicxml contains repeats etc. you can "unfold" it to several different versions using partitura : 
```
score_variants = pt.score.make_score_variants(part) # score variants only contains the structure of the variants, not full parts
part_versions_with_updated_ids = [pt.utils.update_note_ids_after_unfolding(score_variant.create_variant_part()) for score_variant in score_variants]
```
We also have a web-based tool to visualize and check your created alignments: https://sildater.github.io/parangonada/

With these three objects, we can now export 
```
partitura.io.save_match(alignment, performance, score, "my_very_own_match.match")
```
Happy Matching!

**Comment by Alf-Z-SymphoMe** [comment]
Hi, could you please add more instructions on how to create a .match file from scratch? I might be wrong, but the step of creating an alignment seems to require a match file:
```matcher = pa.AutomaticNoteMatcher()
pred_alignment = matcher(sna_match, 
                        pna_match,
                        verbose_time=True)

# compute f-score and print the results
pa.print_fscore_alignments(pred_alignment, groundtruth_alignment)
```
where
```
perf_match, groundtruth_alignment, score_match = pt.load_match(
    filename= pa.EXAMPLE, # 
    create_score=True
)

pna_match = perf_match[0].note_array()
sna_match = score_match[0].note_array()
```
and ```pa.EXAMPLE``` would be the match file.
 Thanks

**Reply by sildater** [reply]
Hi @Alf-Z-SymphoMe 
Thanks for the question! Yes, in the example we start with an existing match file to demonstrate it. The code snippets you posted are from the [parangonar](https://github.com/sildater/parangonar) library, which we'll use to create alignments. If you have a midi performance and a score musicxml file, you can align them like so:
```
import partitura as pt
import parangonar as pa
performance = pt.load_performance_midi("path/to/performance.mid")
score = pt.load_musicxml("path/to/score.musicxml" )
pna = performance[0].note_array()
sna = score[0].note_array(include_grace_notes=True)
matcher = pa.DualDTWNoteMatcher() # I recommend using this matcher by default
predicted_alignment = matcher(sna, pna, process_ornaments=True, score_part=score[0]) 
```
then you can save all three objects (performance, score , and alignment) together in a match file:
```
pt.save_match(predicted_alignment ,
                    performance,
                    score,
                    "path/where/you/want/your/matchfile.match,
                    performer='performer_name', # set these metadata strings if you like
                    composer='composer_name',
                    piece='piece_name',
                    assume_unfolded = True) 
```
This should give you a match file. A couple of caveats:
- if your musicxml has no note IDs, create them using the flag `force_note_ids = True` in `pt.load_musicxml`
- if your score contains multiple parts, merge the, before alignment `merged_part = pt.score.merge_parts(score.parts)`
- if your score contains repeats which are played in the performance, you need to unfold the score to the right structure. This is not trivial, but I can guide you through it should that be the case.

I hope this helps and feel free to ask if anything is unclear!

**Reply by sildater** [reply]
ps: code is off the top of my head, let me know if it doesn't run so I can correct it!

**Comment by Alf-Z-SymphoMe** [comment]
Hi. I managed to create a .match file and an alignment plot. Thanks for your help
